jedi/inference/references.py

```diff
@@ -140,10 +144,15 @@ def find_references(module_context, tree_name):
             search_name,
         )
 
+    gotos = set()
+    if not all_scopes:
+        for name in found_names_dct.values():
+            gotos |= set(name.goto())
+
     non_matching_reference_maps = {}
     for module_context in potential_modules:
         for name_leaf in module_context.tree_node.get_used_names().get(search_name, []):
-            new = _dictionarize(_find_names(module_context, name_leaf))
+            new = _dictionarize(_find_names(module_context, name_leaf, all_scopes=all_scopes))
```

Comment:
I think this all_scopes is also not needed and with that gone, you can probably remove quite a lot of code that modifies `goto`.